### DISCUSSION TITLE
Cursor Tracker: Value & Saturation

### DISCUSSION THREAD

**Discussion by markbehm71** [wojtryb/Shortcut-Composer#63] [Ideas]
The idea: **Vertical movement adjusts foregroundColor value, horizontal adjusts foregroundColor saturation.** Like many XY color palettes. 

I would not have thought this valuable until adapting to it in photoshop w their floating color palette "HUD". They center tha palette on your current Val/Sat rather than just palette-center-on-cursor like most apps.  **It's remarkable how quickly and intuitively you can tweak your current color given holding a key down and dragging a bit up and to one side**. I don't use Hue like this but I could see how someone might like that as an option or under perhaps under a different shortcut.

I have a temperamental Autohotkey script that does this krita's built in hotkeys - but not well and with no visual feedback. Cursor tracker seems to be a perfect solution. What do you think?

**Comment by wojtryb** [comment]
Hi, and thanks for your idea.

Something like that should be fully doable, as krita allows to read and change color in plugins. Expanding this plugin to do so, should be relatively easy.

There are two problems to consider though:

1. There would be no color preview, so you'd need to have a usual color selector docker open somewhere (which is not ideal in canvas-only mode). Writing the color preview makes the task significantly harder.
2. There are already good alternatives in krita like "Show color selector (shift+i)" or "Show shade selector (shift+m)", and I'm not sure my solution would be better than those.

That's why I haven't implemented it yet.

**Reply by sebsepia2** [reply]
Hello! I would also be very interested in this feature. I don't think it needs a color preview to be useful. It would enable me to tweak a color faster than by invoking any color picker, and I can look at a color selector already open, or the color selector in the toolbar, to see the color I am editing. Furthermore if it includes options to tweak R,G,B, LaB or anything else it means I can change the color in a way that is not immediately available with the conventional color wheel. So I can work with a simpler UI without needing a bunch of sliders constantly open. Obviously having a color preview would be sweet, but even without it I would absolutely love this feature. Thanks for considering :)

**Reply by wojtryb** [reply]
Hi,

I've made it work trying to keep this as simple as possible. In `shortcut_composer/actions.py` you need to create a controller for each color channel you want to edit. Thats how it looks for the Red Channel:

![Screenshot_20250419_095830](https://github.com/user-attachments/assets/9ee97eef-a4b0-4797-a827-c0b781606788)

`get_value` must return a value of the red channel (I use the range 0-100, but you could keep the 0-1 or use 0-255, as long you have the correct STEP and action sensitivity).

`set_value` must change the foreground color, editing only the controlled channel with value received in method argument.

For Green and Blue channel controllers you just need to change the index from 0 to 1 and 2 respectively. For L, a, b, H, S, V you either need to figure out how it works in [ManagedColor class](https://api.kde.org/krita/html/classManagedColor.html#a428fe292713b264af5112ae98d74c29c), or use the math to convert between RGB and LAB/HSV.

In `create_actions` you just create a CursorTracker, and define the same action in `actions.action` file.

`shortcut_composer/actions.py`
```python
from composer_utils.label import LabelText
from core_components.controller_base import NumericController
from core_components.controllers.view_controllers import ViewBasedController


class RedChannelController(ViewBasedController, NumericController):

    TYPE = int
    DEFAULT_VALUE: int = 0
    MIN_VALUE = 0
    MAX_VALUE = 100
    STEP = 1
    WRAPPING = False
    ADAPTIVE = True

    def get_value(self) -> float:
        color = self.view.view.foregroundColor()
        return color.components()[0] * 100

    def set_value(self, value: float) -> None:
        color = self.view.view.foregroundColor()
        components = color.components()
        components[0] = value / 100
        color.setComponents(components)
        self.view.view.setForeGroundColor(color)

    def get_label(self, value: float) -> LabelText:
        return LabelText(self.get_pretty_name(value))

    def get_pretty_name(self, value: float) -> str:
        return f"{round(value)}"


def create_actions() -> list[templates.RawInstructions]: return [

    templates.CursorTracker(
        name="Scroll color",
        instructions=[],
        horizontal_slider=Slider(
            controller=RedChannelController(),
            values=Range(0, 100),
            deadzone=100,
        ),
    )
```

`shortcut_composer/actions.action`
```xml
<Action name="Scroll color">
<activationFlags>1</activationFlags>
</Action>
```

**Reply by sebsepia2** [reply]
Holy smokes, thank you so much! I wasnt expecting such a quick and detailed answer. I tried it and it works flawlessly for the RGB values! it is super satisfying.

I tried to get the script working for HSV, and I am kinda getting there. This is kinda janky but it works.. as long as no HSV values are at 0. then it stops working, until i manually set back the values to >1. any idea why that would be?

 `actions.py`
```from PyQt5.QtGui import QColor
from krita import ManagedColor

class SaturationController(ViewBasedController, NumericController):

    TYPE = int
    DEFAULT_VALUE: int = 0
    MIN_VALUE = 0
    MAX_VALUE = 100
    STEP = 1
    WRAPPING = False
    ADAPTIVE = True

    def get_value(self) -> float:
        val = Krita.instance().activeWindow().activeView()
        color = self.view.view.foregroundColor().colorForCanvas(val.canvas())
        return color.getHsvF()[1] * 100

    def set_value(self, value: float) -> None:
        val = Krita.instance().activeWindow().activeView()
        color = self.view.view.foregroundColor().colorForCanvas(val.canvas())
        components = list(color.getHsvF())
        h = components[0]
        s = components[1] = value / 100
        v = components[2]
        color.setHsvF(h,s,v)
        self.view.view.setForeGroundColor(ManagedColor.fromQColor(color, val.canvas()))

    def get_label(self, value: float) -> LabelText:
        return LabelText(self.get_pretty_name(value))

    def get_pretty_name(self, value: float) -> str:
        return f"{round(value)}"
 ```
`actions.action`
```
<Action name="Scroll Saturation">
<activationFlags>1</activationFlags>
</Action> 
```
Thanks again, your script is amazing!

**Reply by wojtryb** [reply]
It's great that you made it work for RGB, and found a way to transform between HSV and RGB.

The issue you talk about most likely comes from the fact that the QColor class seems to set hue to -1 when saturation is 0 (which I believe means hue is undefined). I think that krita treates ManagedColor with hue of -1 invalid and refuses to set it.
At least that is my guess from printing the HSV values in terminal.

Another issue I noticed was some sort of drift of hue while using the slider - it can come from rounding errors.

The fastest way I found to fix that was to store hue and value (channels that are not meant to be changed) inside the object each time. Then I use those, ignoring the ones that come from krita.

I also do not store the hue, when its value is invalid (the if statement). That fixes the problem you mentioned.

Note that this solution will work well only for the CursorTrackers, and may not work well with PieMenus or other action types.

Also note that you can use both horizontal and vertical sliders on one action (see other existing CursorTrackers), and you may need set WRAPPING to True on HueController to make ir switch between 0 and 360 seemlessly.

Good luck :)

```python
class SaturationController(ViewBasedController, NumericController):

    TYPE = int
    DEFAULT_VALUE: int = 0
    MIN_VALUE = 0
    MAX_VALUE = 100
    STEP = 1
    WRAPPING = False
    ADAPTIVE = True

    def __init__(self) -> None:
        self.hue = 0.0
        self.value = 0.0

    def get_value(self) -> float:
        view = self.view.view
        color = view.foregroundColor().colorForCanvas(view.canvas())
        components = color.getHsvF()
        if components[0] >= 0:
            self.hue = components[0]
        self.value = components[2]
        return components[1] * 100

    def set_value(self, value: float) -> None:
        color = QColor()
        color.setHsvF(
            self.hue,
            value / 100,
            self.value)
        self.view.view.setForeGroundColor(
            ManagedColor.fromQColor(color, self.view.view.canvas()))

    def get_label(self, value: float) -> LabelText:
        return LabelText(self.get_pretty_name(value))

    def get_pretty_name(self, value: float) -> str:
        return f"{round(value)}"
```

**Reply by sebsepia2** [reply]
Thanks for answering my many questions! 
Your method works for making the hue stable, but saturation still seems to be intimately tied to the value: the lower the value, the bigger the jumps in saturation when trying to change it. and if value is at 0, saturation won't change at all. Changing the method from HSV to HSL doesnt help, invoking the saturation with hslSaturationF() neither. On paper it seems with your fix of the script the brightness and saturation values should be separated, but theyre still acting jointly.

**Reply by wojtryb** [reply]
Sorry there was no answer from me lately.
I checked a couple of different solutions for your HSV problems but nothing seemed to work.
I've tried using both setHsv and setHsvF with different ranges (0-1 and 0-100), but the issue persists and is the most visible on low saturation.
I've also tried using different conversion functions found on the internet, but the effect is still the same.
I think this is a problem with krita's ManagedColor class, which does not support HSV directly (but it does support Lab, so you could check it), and I don't feel like reading its source code to understand how exactly it works.
Anyway, the last code I provided seems to be mostly usable for personal use.

**Reply by sebsepia2** [reply]
thanks for the update and for looking into it :) TIL krita handles colors in a funny way. Even with that small issue I find using the plugin to change colors on the fly very useful! absolutely understandable that you dont wanna look deep in the bowels of krita for this.
